File: 01/01.2/main.py
INPUT_FILE="/Users/twodarek/personal/advent-of-code-2024/01/01.2/input.txt"

def readInput():
    listA = []
    mapB = {}

    f = open(INPUT_FILE, "r")

    for line in f:
        tmp = line.split()
        listA.append(int(tmp[0]))
        intTmpOne = int(tmp[1])
        if intTmpOne in mapB:
            mapB[intTmpOne] = mapB[intTmpOne] + 1
        else:
            mapB[intTmpOne] = 1

    f.close()
    return listA, mapB


def calcSimilarity(listA, mapB):
    resultant = 0
    for item in listA:
        if item in mapB:
            resultant += item*mapB[item]
    return resultant


def main():
    listA, mapB = readInput()
    similarity = calcSimilarity(listA, mapB)
    print("Similarity score: {0}".format(similarity))


if __name__ == "__main__":
    main()

File: 01/01.2/main.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"strconv"
	"strings"
)

func main() {
	inputFileName := "./input.txt"

	inputFile, err := os.Open(inputFileName)
	if err != nil {
		log.Fatalf("Error: unable to open input file: %s", err)
	}
	defer func(inputFile *os.File) {
		err := inputFile.Close()
		if err != nil {
			log.Printf("Unable to close inputFile: %s\n", err)
		}
	}(inputFile)

	listA := make([]int, 0)
	mapB := make(map[int]int)
	readInput(inputFile, listA, mapB)

}

func readInput(inputFile *os.File, listA []int, mapB map[int]int) {
	scanner := bufio.NewScanner(inputFile)
	for scanner.Scan() {
		line := strings.Fields(scanner.Text())
		if len(line) < 2 {
			log.Printf("Bad input line: %+v\n", line)
			continue
		}
		addStrToArray(line[0], &listA)
		addStrToMap(line[1], mapB)
	}
	fmt.Printf("listA: %+v\n", listA)
	fmt.Printf("mapB: %+v\n", mapB)
	simScore := calculateSimilarity(listA, mapB)
	fmt.Printf("Similarity Score: %d\n", simScore)
}

func calculateSimilarity(listA []int, mapB map[int]int) int {
	simScore := 0
	for _, val := range listA {
		if mapVal, exists := mapB[val]; exists {
			simScore += val * mapVal
		}
	}

	return simScore
}

func addStrToMap(num string, mapInts map[int]int) {
	intNum, err := strconv.Atoi(num)
	if err != nil {
		log.Fatalf("Unable to convert string %s to int: %s", intNum, err)
	}
	if val, exists := mapInts[intNum]; exists {
		mapInts[intNum] = val + 1
	} else {
		mapInts[intNum] = 1
	}
}

func addStrToArray(num string, listInts *[]int) {
	intNum, err := strconv.Atoi(num)
	if err != nil {
		log.Fatalf("Unable to convert string %s to int: %s", intNum, err)
	}
	*listInts = append(*listInts, intNum)
}
